# **How grocery purchases affect Eduational Attainment**

The goal of this project is to understand how customers in London's boroughs make purchases, and how that affects educational attainment within the borough. With just under 3000 locations in the UK, Tesco is a multinational retailer of groceries and miscellaneous goods with around a 28% market share in the UK. It is an ideal source for analyzing customer buying habits because of its wide reach. We will be using The Tesco Grocery 1.0 Dataset for this project. 

[Link for the Tesco Grocery 1.0 Dataset](https://figshare.com/articles/dataset/Area-level_grocery_purchases/7796666?backTo=/collections/Tesco_Grocery_1_0/4769354)

The Tesco Grocery 1.0 Dataset, which was collected from Tesco Clubcard customers, serves as the basis for the analysis. This dataset, which combines demographic and geographic data with transactional data from Tesco locations for 2015, gives a broad insight into grocery purchasing habits. In order to compare the patterns  in grocery consumption in different regions of London, the data for this study has been gathered at the borough level.



## **Preprocessing**

In [180]:
# importing necessary libraries
import pandas as pd
import geopandas as gpd

In [181]:
# loading the Tesco Grocery 1.0 dataset
grocery_data = pd.read_csv('/Users/s16teen/Downloads/Work/UOB/SEM2/ADS2/Tesco_Grocery_and_Educational_Attainment/Data/year_borough_grocery.csv')

grocery_data.head(1)

area_id     weight  weight_perc2.5  weight_perc25  weight_perc50  \
0  E09000001  323.17475            35.0          150.0          250.0   

   weight_perc75  weight_perc97.5  weight_std  weight_ci95     volume  ...  \
0          400.0           1000.0  323.692596      1.00152  98.842963  ...   

   man_day  population    male  female  age_0_17  age_18_64  age_65+  \
0   103883      6687.0  3697.0  2990.0    1018.0     4349.0   1320.0   

     avg_age  area_sq_km  people_per_sq_km  
0  43.910124         2.9       2305.862069  

[1 rows x 202 columns]

In [182]:
# loading the 'Statistical GIS Boundary Files for London' dataset
map_df = gpd.read_file("/Users/s16teen/Downloads/Work/UOB/SEM2/ADS2/Tesco_Grocery_and_Educational_Attainment/Data/Chorpleth Datasets/statistical-gis-boundaries-london/ESRI/London_Borough_Excluding_MHW.shp")

map_df.head(1)

NAME   GSS_CODE  HECTARES  NONLD_AREA ONS_INNER SUB_2009  \
0  Kingston upon Thames  E09000021  3726.117         0.0         F     None   

  SUB_2006                                           geometry  
0     None  POLYGON ((516401.6 160201.8, 516407.3 160210.5...

In [183]:
# converting the datasets to pandas dataframes
grocery_data = pd.DataFrame(grocery_data)

In [184]:
# adding the names of the boroughs from the 'Statistical GIS Boundary Files for London' dataset
grocery_data = grocery_data.set_index('area_id').join(map_df.set_index('GSS_CODE')[['NAME']])

grocery_data = grocery_data.reset_index()

In [185]:
# displaying the first 5 rows of the dataset
grocery_data.head()

area_id      weight  weight_perc2.5  weight_perc25  weight_perc50  \
0  E09000001  323.174750            35.0          150.0          250.0   
1  E09000002  421.304125            32.5          165.6          300.0   
2  E09000003  407.170491            40.0          180.0          320.0   
3  E09000004  359.544126            30.0          144.0          255.0   
4  E09000005  434.324141            40.0          175.0          325.0   

   weight_perc75  weight_perc97.5  weight_std  weight_ci95      volume  ...  \
0          400.0           1000.0  323.692596     1.001520   98.842963  ...   
1          500.0           1500.0  597.994271     0.828993  122.341294  ...   
2          500.0           1160.0  443.644324     0.171615  114.933771  ...   
3          450.0           1000.0  476.576945     0.819748  106.446426  ...   
4          500.0           1500.0  616.087316     0.354492  128.930369  ...   

   population      male    female  age_0_17  age_18_64  age_65+    avg_age  \
0      6687.0    3697.0    2990.0    1018.0     4349.0   1320.0  43.910124   
1    203101.0   99442.0  103659.0   60281.0   123180.0  19640.0  33.017321   
2    378778.0  186357.0  192421.0   88754.0   237036.0  52988.0  37.195582   
3    242387.0  116814.0  125573.0   56028.0   146237.0  40122.0  38.880749   
4    323443.0  163939.0  159504.0   75033.0   211885.0  36525.0  35.766416   

   area_sq_km  people_per_sq_km                  NAME  
0        2.90       2305.862069        City of London  
1       36.09       5627.625381  Barking and Dagenham  
2       86.76       4365.813739                Barnet  
3       60.58       4001.105976                Bexley  
4       43.28       7473.267098                 Brent  

[5 rows x 203 columns]

## **Task 1: Describe the Dataset**

In this task you need to describe the dataset. It contains many fields. A good description will find ways
to group fields together to summarise the content rather than simply list all the fields. You need to
explain the value of the data rather than only provide basic description. You should also address any
assumptions or limitations that need to be considered which may affect how the data can or should be used.

### **Basic Summary of the Tesco Dataset**

In [186]:
# finding the shape of the dataset
grocery_data.shape

(33, 203)

In [187]:
# finding if there are any missing values in the dataset
grocery_data.isnull().sum().sum()

0

In [188]:
# displaying the first few rows of the dataset
grocery_data.head()

area_id      weight  weight_perc2.5  weight_perc25  weight_perc50  \
0  E09000001  323.174750            35.0          150.0          250.0   
1  E09000002  421.304125            32.5          165.6          300.0   
2  E09000003  407.170491            40.0          180.0          320.0   
3  E09000004  359.544126            30.0          144.0          255.0   
4  E09000005  434.324141            40.0          175.0          325.0   

   weight_perc75  weight_perc97.5  weight_std  weight_ci95      volume  ...  \
0          400.0           1000.0  323.692596     1.001520   98.842963  ...   
1          500.0           1500.0  597.994271     0.828993  122.341294  ...   
2          500.0           1160.0  443.644324     0.171615  114.933771  ...   
3          450.0           1000.0  476.576945     0.819748  106.446426  ...   
4          500.0           1500.0  616.087316     0.354492  128.930369  ...   

   population      male    female  age_0_17  age_18_64  age_65+    avg_age  \
0      6687.0    3697.0    2990.0    1018.0     4349.0   1320.0  43.910124   
1    203101.0   99442.0  103659.0   60281.0   123180.0  19640.0  33.017321   
2    378778.0  186357.0  192421.0   88754.0   237036.0  52988.0  37.195582   
3    242387.0  116814.0  125573.0   56028.0   146237.0  40122.0  38.880749   
4    323443.0  163939.0  159504.0   75033.0   211885.0  36525.0  35.766416   

   area_sq_km  people_per_sq_km                  NAME  
0        2.90       2305.862069        City of London  
1       36.09       5627.625381  Barking and Dagenham  
2       86.76       4365.813739                Barnet  
3       60.58       4001.105976                Bexley  
4       43.28       7473.267098                 Brent  

[5 rows x 203 columns]

In [189]:
# Basic statistics of the dataset
grocery_data.describe()

weight  weight_perc2.5  weight_perc25  weight_perc50  \
count   33.000000       33.000000      33.000000      33.000000   
mean   373.898533       35.954545     159.969697     284.581818   
std     40.721303        4.051164      16.896088      32.326396   
min    295.330271       28.000000     120.000000     215.000000   
25%    348.420414       32.500000     150.000000     250.000000   
50%    374.737951       37.000000     165.000000     300.000000   
75%    405.993196       40.000000     175.000000     305.000000   
max    450.349903       40.000000     180.000000     325.000000   

       weight_perc75  weight_perc97.5  weight_std  weight_ci95      volume  \
count      33.000000        33.000000   33.000000    33.000000   33.000000   
mean      469.212121      1113.939394  434.344901     0.422137  109.808642   
std        39.005414       175.605015   96.926163     0.241992   10.405026   
min       400.000000      1000.000000  300.762891     0.171615   95.006632   
25%       450.000000      1000.000000  359.841786     0.271506  102.882176   
50%       500.000000      1000.000000  425.267620     0.323224  106.446426   
75%       500.000000      1160.000000  486.498044     0.487412  114.933771   
max       500.000000      1500.000000  718.157462     1.001520  136.703561   

       volume_perc2.5  ...       man_day     population           male  \
count       33.000000  ...  3.300000e+01      33.000000      33.000000   
mean        11.848485  ...  1.641727e+06  262634.242424  130472.909091   
std          1.598851  ...  1.139058e+06   73791.635356   36906.723911   
min         10.100000  ...  1.038830e+05    6687.000000    3697.000000   
25%         11.000000  ...  7.342320e+05  227507.000000  113683.000000   
50%         11.000000  ...  1.485336e+06  268626.000000  134797.000000   
75%         12.500000  ...  2.612931e+06  319477.000000  154118.000000   
max         15.000000  ...  4.131210e+06  380070.000000  186357.000000   

              female      age_0_17      age_18_64       age_65+    avg_age  \
count      33.000000     33.000000      33.000000     33.000000  33.000000   
mean   132161.333333  59143.484848  173158.060606  30332.696970  36.253457   
std     37158.195707  19011.948712   49417.623844  11368.467277   2.498356   
min      2990.000000   1018.000000    4349.000000   1320.000000  31.514726   
25%    113824.000000  46242.000000  146237.000000  23417.000000  34.858641   
50%    135020.000000  61120.000000  179854.000000  28265.000000  35.934276   
75%    159504.000000  70644.000000  211354.000000  36907.000000  37.195582   
max    195765.000000  93166.000000  237511.000000  56804.000000  43.910124   

       area_sq_km  people_per_sq_km  
count   33.000000         33.000000  
mean    47.650909       7286.526000  
std     32.714349       3887.905684  
min      2.900000       2166.664447  
25%     26.820000       4390.319972  
50%     37.640000       5793.247521  
75%     56.380000      10687.248787  
max    150.140000      15268.926174  

[8 rows x 201 columns]

<mark>**Key Highlights from the dataset**</mark>

- The dataset contains information about each Borough. (We know theres 33 Boroughs and there are 33 rows in the datset)
- There are 202 columns, which will be grouped up and summarized to simplify the task.
- There are no missing values in the dataset.
- The dataset is thorough and includes summary statistics of each variable like the percentile of each variable, the standard deviation and the 95% confidence intervals.
- The average product weight is 374g.
- The average product volume is 110mL.
- Boroughs vary in population from 6,687 to 380,070.
- The population density ranges from 2,166 to 15,269 people/km².

<mark>**We can group up the columns and summarize each group**</mark>

*The main groups of focus are:*
- Nutrients 
- Product Charateristics
- Product Categories
- Purchase Behaviours
- Demographic Information
- Geographic Information

### **Nutrient information**

The dataset contains columns for the weight, amount of energy and entropy of each nutrient in the average product.

These nutrients are:
- Carbs: Total carbohydrates. (includes sugars)
- Sugar: Natural and added sugars.
- Fats: Total fat content. (includes saturated fat)
- Saturated Fats: Total saturated fat content
- Protein: Amount of protein.
- Fibre: Total fiber content.
- Salt: Total Salt content.
- Alcohol: Total alcohol content

These columns provides insight into the composition of food products purchased in different boroughs. This can help measure nutritional quality in different boroughs and the potential health implications.

***Lets try to find some insights for the Nutrients***

In [190]:
# finding the average of the nutrients columns
nutrients_cols = ['carb', 'sugar', 'fat', 'saturate', 'protein', 'fibre', 'salt', 'alcohol']
nutrient_avg = grocery_data[nutrients_cols].mean()

nutrient_avg

carb        17.967103
sugar       10.018118
fat          8.964480
saturate     3.527514
protein      5.299996
fibre        1.618842
salt         0.580529
alcohol      0.242141
dtype: float64

In [191]:
# finding which borough has the highest and lowest average protein content
print("Highest Protein Consumption: ", grocery_data[grocery_data['protein'] == grocery_data['protein'].max()]['NAME'].values[0])
print("Lowest Protein Consumption: ", grocery_data[grocery_data['protein'] == grocery_data['protein'].min()]['NAME'].values[0])

Highest Protein Consumption:  Southwark
Lowest Protein Consumption:  Newham


In [192]:
# finding which borough has the highest and lowest average sugar content
print("Highest Sugar Consumption: ", grocery_data[grocery_data['sugar'] == grocery_data['sugar'].max()]['NAME'].values[0])
print("Lowest Sugar Consumption: ", grocery_data[grocery_data['sugar'] == grocery_data['sugar'].min()]['NAME'].values[0])

Highest Sugar Consumption:  Sutton
Lowest Sugar Consumption:  Kensington and Chelsea


### **Product Charateristics**

The dataset has columns for the weight, volume, amount of calories and concentration of calories in the average product, measured in kcals and kcals/gram respectively. This information can be useful to find boroughs with high-energy and low-energy diets.

***Lets try to find some insights for the Energy***

In [193]:
# finding which borough has the highest and lowest calorie values
print("Highest Calorie Consumption: ", grocery_data[grocery_data['energy_tot'] == grocery_data['energy_tot'].max()]['NAME'].values[0])
print("Lowest Calorie Consumption: ", grocery_data[grocery_data['energy_tot'] == grocery_data['energy_tot'].min()]['NAME'].values[0])

Highest Calorie Consumption:  Havering
Lowest Calorie Consumption:  City of London


In [194]:
# finding which borough has the highest and lowest weight values
print("Highest Weight Consumption: ", grocery_data[grocery_data['weight'] == grocery_data['weight'].max()]['NAME'].values[0])
print("Lowest Weight Consumption: ", grocery_data[grocery_data['weight'] == grocery_data['weight'].min()]['NAME'].values[0])

Highest Weight Consumption:  Newham
Lowest Weight Consumption:  Sutton


In [195]:
# finding which borough has the highest and lowest volume values
print("Highest Volume Consumption: ", grocery_data[grocery_data['volume'] == grocery_data['volume'].max()]['NAME'].values[0])
print("Lowest Volume Consumption: ", grocery_data[grocery_data['volume'] == grocery_data['volume'].min()]['NAME'].values[0])

Highest Volume Consumption:  Newham
Lowest Volume Consumption:  Kingston upon Thames


<mark>**We can see from this that the weight and volume of the average product being higher doesn't necessarily correlate to more calories in the average product.**</mark>

### **Product Categories**

The dataset records the proportion of different food and beverage types consumed.

The different categories are:
- Beer
- Dairy
- Eggs
- Fats and Oils
- Fish
- Fruits and Vegetables
- Grains
- Red Meat
- Poultry
- Readymade Foods
- Sauces
- Soft Drinks
- Spirits
- Sweets
- Tea and Coffee
- Water
- Wine

These categories are important as we can see which boroughs buy alot of products that are either 'junk food' or 'healthy food', and this can help us find healthy boroughs.

***Lets try to find some insights for the Product Categories***

In [196]:
# finding boroughs with most and least junk food consumption
junk_food_cols = ['f_beer', 'f_soft_drinks', 'f_spirits', 'f_sweets', 'f_wine']
grocery_data['junk_food'] = grocery_data[junk_food_cols].sum(axis=1)

print("Borough with most junk food consumption: ", grocery_data[grocery_data['junk_food'] == grocery_data['junk_food'].max()]['NAME'].values[0])
print("Borough with least junk food consumption: ", grocery_data[grocery_data['junk_food'] == grocery_data['junk_food'].min()]['NAME'].values[0])

Borough with most junk food consumption:  Sutton
Borough with least junk food consumption:  Kensington and Chelsea


In [197]:
# finding boroughs with most and least healthy food consumption
healthy_food_cols = ['f_eggs', 'f_fruit_veg', 'f_fish', 'f_meat_red', 'f_poultry']
grocery_data['healthy_food'] = grocery_data[healthy_food_cols].sum(axis=1)

print("Borough with most healthy food consumption: ", grocery_data[grocery_data['healthy_food'] == grocery_data['healthy_food'].max()]['NAME'].values[0])
print("Borough with least healthy food consumption: ", grocery_data[grocery_data['healthy_food'] == grocery_data['healthy_food'].min()]['NAME'].values[0])

Borough with most healthy food consumption:  Kensington and Chelsea
Borough with least healthy food consumption:  Sutton


<mark>**Something interesting to note is that Kensignton and Chelsea have the highest consumption of foods that are considered healthy and lowest consumption of foods that are considered unhealthy. The same, but opposite, for Sutton which have high junk food consumption and low healthy food consumption**</mark>

### **Purchase Behaviours**

This group includes metrics on the number of transactions, frequency of purchases, and consumer participation rates. The diversity of product weights and transaction days highlights purchasing patterns, which can be useful for market analysis and public health studies.

***Lets try to find some insights for the Purchase Behaviours***

In [198]:
# finding boroughs with the most and least transactions
print("Borough with most transactions: ", grocery_data[grocery_data['num_transactions'] == grocery_data['num_transactions'].max()]['NAME'].values[0])
print("Borough with least transactions: ", grocery_data[grocery_data['num_transactions'] == grocery_data['num_transactions'].min()]['NAME'].values[0])

Borough with most transactions:  Barnet
Borough with least transactions:  City of London


In [199]:
# finding boroughs with the most transactions per day
grocery_data['avg_transactions_per_day'] = grocery_data['num_transactions'] / grocery_data['transaction_days']
grocery_data[['NAME', 'avg_transactions_per_day']].sort_values(by='avg_transactions_per_day', ascending=False).head()

NAME  avg_transactions_per_day
2           Barnet             158974.041096
9          Enfield             136756.597260
30  Waltham Forest             126876.578082
27       Southwark             111131.008219
21         Lambeth              91504.172603

### **Demographic Information**

Population statistics, including age distribution and gender, are given in the dataset and provide a broader context for understanding dietary and healthiness trends. The average age and population structure help correlate food consumption with demographic factors.

***Lets try to find some insights for the Demographic Information***

In [200]:
# finding boroughs with more females than males
grocery_data['fem_ratio'] = grocery_data['female'] / (grocery_data['male'])
female_majority = grocery_data[['NAME', 'fem_ratio']].sort_values(by='fem_ratio', ascending=False).head(5)
print("Boroughs with More Females than Males:\n", female_majority)

Boroughs with More Females than Males:
           NAME  fem_ratio
15    Havering   1.082585
5      Bromley   1.081619
3       Bexley   1.074982
31  Wandsworth   1.072938
7      Croydon   1.062180


In [201]:
# finding the boroughs with the lowest average ages
grocery_data[['NAME', 'avg_age']].sort_values(by='avg_age', ascending=True).head()

NAME    avg_age
29         Tower Hamlets  31.514726
24                Newham  32.038272
11               Hackney  32.973528
1   Barking and Dagenham  33.017321
27             Southwark  34.200656

### **Geographic Information**

Columns for the area size, population density, and customer representativeness are in the dataset and help us get further explainations for the borough shopping trends.

***Lets try to find some insights for the Geographic Information***

In [202]:
# finding most densely populated areas
dense_areas = grocery_data[['NAME', 'people_per_sq_km']].sort_values(by='people_per_sq_km', ascending=False).head(5)
print("Most Densely Populated Boroughs:\n", dense_areas)

Most Densely Populated Boroughs:
                       NAME  people_per_sq_km
18               Islington      15268.926174
29           Tower Hamlets      14877.367089
11                 Hackney      14078.930818
19  Kensington and Chelsea      13095.706028
21                 Lambeth      11958.836689


In [203]:
# finding areas with highest Clubcard participation (better customer data)
high_representativeness = grocery_data[['NAME', 'representativeness_norm']].sort_values(by='representativeness_norm', ascending=False).head(5)
print("Boroughs with Highest Clubcard Representativeness:\n", high_representativeness)

Boroughs with Highest Clubcard Representativeness:
               NAME  representativeness_norm
0   City of London                 1.000000
14          Harrow                 0.747110
2           Barnet                 0.720439
30  Waltham Forest                 0.701978
9          Enfield                 0.679551


### **Limiations**

- Incomplete Coverage of Food Purchases: The dataset does not include purchases from other supermarket chains, or small stores [3].
- Excludes Non-Clubcard & Online Purchases: Information about Tesco customers who do not use a Clubcard or shop online are missing from the dataset [3].
- No Data on Restaurant/Takeaway Food: The dataset does not account for food consumed outside the home, which forms a large part of the average persons diet [3].
- Area Level Consumption: Nutritional values are based on an area-level average rather than individual-level consumption, limiting personal dietary insights [3].
- Low Data Representation in Some Areas: Certain boroughs have less Tescos, making data unreliable for those locations [3].
- The dataset is limited to the year 2015.

### Assumptions

- The dataset assumes that the general shopping behavior for each of the boroughs is represented by the purchasing patterns of Tesco Clubcard owners.
- The average product reflects actual dietary and health patterns.
- All food categories are equally likely to be purchased.
- The data collected is accurate.

### Biases

- Self-Selection Bias: The dataset includes only Tesco Clubcard holders, meaning the sample is not random but self-selected [3].
- Demographic & Socio-Economic Bias: The data does not fully represent the population in terms of age, income, or social background [3].
- Geographic Bias: Some areas have higher Tesco store concentration (North London), leading to unequal data representation across boroughs [3].
- Purchase Behavior Bias: The dataset only captures Tesco shopping habits, excluding preferences of people who primarily shop elsewhere [3].

## Task 2: Visualise the Dataset
In this task you will explore the dataset and use one or more appropriate visualisations to extract two
useful insights from it. You are free to choose both the topics for exploration. The visualisations should
conform with the best practice you studied in Semester 1. You are free to create programs and use any
libraries or other tools to assist you in this task. You are also encouraged to make use of the skills you
have developed in other units such as statistical methods or machine learning approaches where you
consider these appropriate.

### Visualization 1: Healthiest boroughs in London.

Foods that are higher in protein, fiber, vitamins, and minerals and lower in saturated fats, sugars, and sodium are considered nutrient-rich foods. These foods may also contain other bioactive substances that seem to have positive health impacts. Nutrient-rich foods are typically regarded as a component of a healthy diet because of their nutritional and bioactive makeup [4].

## References

1. Aiello, Luca Maria (2020). Area-level grocery purchases. figshare. Dataset. https://doi.org/10.6084/m9.figshare.7796666.v1

2. Greater London Authority (GLA), (2014) 'Statistical GIS Boundary Files for London'. Available at: https://data.london.gov.uk/dataset/statistical-gis-boundary-files-london.

3. Aiello, L.M., Quercia, D., Schifanella, R. et al. Tesco Grocery 1.0, a large-scale dataset of grocery purchases in London. Sci Data 7, 57 (2020). https://doi.org/10.1038/s41597-020-0397-7

4. González-Palacios, S. and Fonollá, J., 2023. Advances in Nutrient-Rich Foods for a Healthy Diet. Foods, 12(15), p.2946.